lib/widgets/post/post_media.dart

```diff
@@ -1,32 +1,72 @@
 import 'package:flutter/material.dart';
+import 'package:flutter_hooks/flutter_hooks.dart';
+import 'package:logging/logging.dart';
+import 'package:path/path.dart';
 
 import '../../util/observer_consumers.dart';
 import '../cached_network_image.dart';
 import '../fullscreenable_image.dart';
+import '../media_providers/liftoff_media_provider.dart';
+import '../media_providers/redgif_provider.dart';
 import 'post_store.dart';
+import 'post_video.dart' as video;
+
+final _logger = Logger('postMedia');
 
 /// assembles image
-class PostMedia extends StatelessWidget {
+class PostMedia extends HookWidget {
   const PostMedia();
 
+  static const redgifProvider = RedgifProvider();
+  static const mp4Provider = MP4MediaProvider();
+
   @override
   Widget build(BuildContext context) {
     return ObserverBuilder<PostStore>(
       builder: (context, store) {
         final post = store.postView.post;
-        if (!store.hasMedia) return const SizedBox();
-
-        final url = post.url!; // hasMedia returns false if url is null
-
-        return FullscreenableImage(
-          url: url,
-          child: CachedNetworkImage(
-            imageUrl: url,
-            errorBuilder: (_, ___) => const Icon(Icons.warning),
-            loadingBuilder: (context, progress) =>
-                CircularProgressIndicator.adaptive(value: progress?.progress),
-          ),
-        );
+
+        if (!store.hasMedia && post.url == null) return const SizedBox();
+
+        final url =
+            Uri.parse(post.url!); // hasMedia returns false if url is null
+
+        _logger.finer(
+            'MEDIA URL: extension: ${extension(url.path)} host: ${store.urlDomain}');
+
+        LiftoffMediaProvider? provider;
+
+        if (redgifProvider.providesFor(url)) {
+          provider = redgifProvider;
+        } else if (mp4Provider.providesFor(url)) {
+          provider = mp4Provider;
+        }
+
+        if (provider != null) {
+          return FutureBuilder(
```

Comment:
You should investigate using the `useFuture()` hook for this. It will be much easier than all the complex stuff gong on in`PostVideo`.

But that should be in a future PR though - `FutureBuilder` is good enough to go into master. :slightly_smiling_face: